cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/datastore/Database200MigrationTest.java

```diff
@@ -0,0 +1,530 @@
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+package com.cloudant.sync.datastore;
+
+import static org.hamcrest.Matchers.is;
+
+import com.cloudant.mazha.Document;
+import com.cloudant.sync.datastore.migrations.MigrateDatabase100To200;
+import com.cloudant.sync.sqlite.Cursor;
+import com.cloudant.sync.sqlite.SQLDatabase;
+import com.cloudant.sync.sqlite.SQLDatabaseQueue;
+import com.cloudant.sync.sqlite.SQLQueueCallable;
+import com.cloudant.sync.util.DatabaseUtils;
+import com.cloudant.sync.util.TestUtils;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Created by rhys on 15/08/2016.
+ */
+public class Database200MigrationTest {
+
+    private Datastore ds;
+    private DatastoreManager manager;
+    private DocumentRevision rootRevision;
+    private String dir;
+
+    @Before
+    public void setUp() throws Exception {
+        dir = TestUtils.createTempTestingDir(this.getClass().getName());
+        manager =  DatastoreManager.getInstance(dir);
+        ds = manager.openDatastore("complexTree");
+
+        Map<String,String> body = new HashMap<String,String>();
+        body.put("hello","world");
+
+        DocumentRevision rev = new DocumentRevision();
+        rev.setBody(DocumentBodyFactory.create(body));
+        rootRevision = ds.createDocumentFromRevision(rev);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        ds.close();
+        TestUtils.deleteTempTestingDir(dir);
+    }
+
+    /**
+     * Tree:
+     *
+     * abc:
+     *
+     * 1-xxxxxxx
+     *
+     * 1-xxxxxxx
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testDatabaseWithComplexTree() throws Exception {
+
+        // sets up the datastore with 1-x and 1-x revisions (Duplicates)
+
+        final String sql = "INSERT INTO revs (doc_id, parent, current, deleted, available, revid, json) SELECT doc_id, parent, current, deleted, available, revid, json FROM revs where revs.revid = ?";
+        final String updateCurrent = "UPDATE revs SET current=0 where sequence=2";
+
+        ((DatastoreImpl)ds).runOnDbQueue(new SQLQueueCallable<Void>() {
+            @Override
+            public Void call(SQLDatabase db) throws Exception {
+                db.execSQL(sql, new String[]{rootRevision.getRevision()});
+                db.execSQL(updateCurrent, new String[]{});
+                return null;
+            }
+        }).get();
+
+        DocumentRevision rev = new DocumentRevision(rootRevision.id, "1-abc");
+        rev.setBody(rootRevision.getBody());
+        ((DatastoreImpl) ds).forceInsert(rev, "1-abc");
+        runMigration();
+
+
+        int count = revisionCount();
+        Assert.assertThat(count, is(2));
```

Comment:
I think this needs more description about why the expected result is 2.

As a general rule I find having a "should" message on assertions adds clarity when reading test code and in error messages. 
e.g.
`Assert.assertThat("A duplicate revision should be removed", count, is(2))`
results in:
 `A duplicate revision should be removed: Expected 2 was 3`
vs:
 ` Expected 2 was 3`

As an aside it confuses me every time that the hamcrest assertion is in the opposite order to the normal JUnit `expected, actual`!